crates/storage/provider/src/bundle_state/state_reverts.rs

```diff
@@ -78,30 +67,54 @@ impl StateReverts {
 
-                tracing::trace!(target: "provider::reverts", ?address, ?storage, "Writing storage reverts");
+                tracing::trace!(target: "provider::reverts", %address, ?storage, "Writing storage reverts");
                 for (key, value) in StorageRevertsIter::new(storage, wiped_storage) {
-                    storage_changeset_cursor.append_dup(storage_id, StorageEntry { key, value })?;
+                    if should_append_storage {
+                        storage_changeset_cursor
+                            .append_dup(storage_id, StorageEntry { key, value })?;
+                    } else {
+                        if let Some(entry) =
+                            storage_changeset_cursor.seek_by_key_subkey(storage_id, key)?
+                        {
+                            if entry.key == key {
+                                tracing::warn!(target: "provider::reverts", ?storage_id, ?entry, "Overwriting storage changeset entry");
+                                storage_changeset_cursor.delete_current()?;
+                            }
+                        }
+
+                        storage_changeset_cursor.upsert(storage_id, StorageEntry { key, value })?;
+                    }
                 }
             }
         }
 
         // Write account changes
         tracing::trace!(target: "provider::reverts", "Writing account changes");
         let mut account_changeset_cursor = tx.cursor_dup_write::<tables::AccountChangeSets>()?;
+        let should_append_accounts = account_changeset_cursor
+            .last()?
+            .map_or(true, |(block_number, _)| block_number < first_block);
         for (block_index, mut account_block_reverts) in self.0.accounts.into_iter().enumerate() {
             let block_number = first_block + block_index as BlockNumber;
             // Sort accounts by address.
             account_block_reverts.par_sort_by_key(|a| a.0);
 
             for (address, info) in account_block_reverts {
-                if exhaustive_data_for_keys {
+                if should_append_accounts {
                     account_changeset_cursor.append_dup(
                         block_number,
                         AccountBeforeTx { address, info: info.map(into_reth_acc) },
                     )?;
                 } else {
-                    // upsert on dupsort tables will append to subkey. see implementation of
-                    // DbCursorRW::upsert for reth_db::implementation::mdbx::cursor::Cursor<RW, _>
+                    if let Some(entry) =
+                        account_changeset_cursor.seek_by_key_subkey(block_number, address)?
+                    {
+                        if entry.address == address {
+                            tracing::warn!(target: "provider::reverts", block_number, ?entry, "Overwriting account changeset entry");
+                            account_changeset_cursor.delete_current()?;
+                        }
+                    }
```

Comment:
this won't happen. upsert does exactly what we want. it's documented for impl of `DbCursorRW`.

Reply:
this is a foolproof check